File: scripts/wayback-sitemap.py
"""
Fetch archived URLs under albatros.uz/product from the Wayback Machine (CDX API)
for a given year and write them out as a plain sitemap list + XML sitemap.

Usage:
    python scripts/wayback-sitemap.py
    python scripts/wayback-sitemap.py --year 2025 --domain albatros.uz --path /product
"""

import argparse
import json
import sys
import urllib.parse
import urllib.request

CDX_ENDPOINT = "http://web.archive.org/cdx/search/cdx"


def fetch_cdx(domain: str, path: str, year: int) -> list[dict]:
    match_url = f"{domain}{path}*"
    params = {
        "url": match_url,
        "from": f"{year}0101",
        "to": f"{year}1231",
        "output": "json",
        "fl": "original,timestamp,statuscode,mimetype",
        "collapse": "urlkey",
        "filter": "statuscode:200",
    }
    query = urllib.parse.urlencode(params)
    full_url = f"{CDX_ENDPOINT}?{query}"

    req = urllib.request.Request(full_url, headers={"User-Agent": "Mozilla/5.0"})
    with urllib.request.urlopen(req, timeout=60) as resp:
        data = json.loads(resp.read().decode("utf-8"))

    if not data:
        return []

    header, *rows = data
    return [dict(zip(header, row)) for row in rows]


def write_outputs(entries: list[dict], out_prefix: str) -> None:
    urls = sorted({e["original"] for e in entries})

    txt_path = f"{out_prefix}.txt"
    with open(txt_path, "w", encoding="utf-8") as f:
        for u in urls:
            f.write(u + "\n")

    xml_path = f"{out_prefix}.xml"
    with open(xml_path, "w", encoding="utf-8") as f:
        f.write('<?xml version="1.0" encoding="UTF-8"?>\n')
        f.write('<urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9">\n')
        for u in urls:
            f.write(f"  <url><loc>{u}</loc></url>\n")
        f.write("</urlset>\n")

    print(f"Found {len(urls)} unique URLs.")
    print(f"Wrote: {txt_path}")
    print(f"Wrote: {xml_path}")


def main():
    parser = argparse.ArgumentParser(description="Build a sitemap from Wayback Machine snapshots.")
    parser.add_argument("--domain", default="albatros.uz")
    parser.add_argument("--path", default="/product")
    parser.add_argument("--year", type=int, default=2025)
    parser.add_argument("--out", default=None, help="Output file prefix (no extension)")
    args = parser.parse_args()

    out_prefix = args.out or f"wayback-sitemap-{args.domain.replace('.', '_')}-{args.year}"

    print(f"Querying Wayback CDX for {args.domain}{args.path}* in {args.year}...")
    try:
        entries = fetch_cdx(args.domain, args.path, args.year)
    except urllib.error.URLError as e:
        print(f"Request failed: {e}", file=sys.stderr)
        sys.exit(1)

    if not entries:
        print("No archived snapshots found for that domain/path/year.")
        sys.exit(0)

    write_outputs(entries, out_prefix)


if __name__ == "__main__":
    main()

File: scripts/wayback-sitemap.mjs
// Fetch archived URLs under a domain/path from the Wayback Machine (CDX API)
// for a given year and write them out as a plain sitemap list + XML sitemap.
//
// Usage:
//   node scripts/wayback-sitemap.mjs
//   node scripts/wayback-sitemap.mjs --domain albatros.uz --path /product --year 2025

import { writeFile } from "node:fs/promises";

function parseArgs(argv) {
  const args = { domain: "albatros.uz", path: "/product", year: 2025, out: null, from: null, to: null };
  for (let i = 0; i < argv.length; i++) {
    const key = argv[i];
    if (key === "--domain") args.domain = argv[++i];
    else if (key === "--path") args.path = argv[++i];
    else if (key === "--year") args.year = Number(argv[++i]);
    else if (key === "--out") args.out = argv[++i];
    else if (key === "--from") args.from = argv[++i];
    else if (key === "--to") args.to = argv[++i];
  }
  return args;
}

async function fetchCdx(domain, path, year, fromOverride, toOverride) {
  const matchUrl = `${domain}${path}*`;
  const params = new URLSearchParams({
    url: matchUrl,
    from: fromOverride ?? `${year}0101`,
    to: toOverride ?? `${year}1231`,
    output: "json",
    fl: "original,timestamp,statuscode,mimetype",
    collapse: "urlkey",
    filter: "statuscode:200",
  });

  const fullUrl = `http://web.archive.org/cdx/search/cdx?${params.toString()}`;
  const res = await fetch(fullUrl, { headers: { "User-Agent": "Mozilla/5.0" } });
  if (!res.ok) {
    throw new Error(`CDX request failed: ${res.status} ${res.statusText}`);
  }
  const data = await res.json();
  if (!data || data.length === 0) return [];

  const [header, ...rows] = data;
  return rows.map((row) => Object.fromEntries(header.map((h, i) => [h, row[i]])));
}

async function writeOutputs(entries, outPrefix) {
  const urls = [...new Set(entries.map((e) => e.original))].sort();

  const txtPath = `${outPrefix}.txt`;
  await writeFile(txtPath, urls.join("\n") + "\n", "utf-8");

  const xmlPath = `${outPrefix}.xml`;
  const body = urls.map((u) => `  <url><loc>${u}</loc></url>`).join("\n");
  const xml = `<?xml version="1.0" encoding="UTF-8"?>\n<urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9">\n${body}\n</urlset>\n`;
  await writeFile(xmlPath, xml, "utf-8");

  console.log(`Found ${urls.length} unique URLs.`);
  console.log(`Wrote: ${txtPath}`);
  console.log(`Wrote: ${xmlPath}`);
}

async function main() {
  const args = parseArgs(process.argv.slice(2));
  const outPrefix = args.out ?? `wayback-sitemap-${args.domain.replace(/\./g, "_")}-${args.year}`;

  console.log(`Querying Wayback CDX for ${args.domain}${args.path}* in ${args.year}...`);
  const entries = await fetchCdx(args.domain, args.path, args.year, args.from, args.to);

  if (entries.length === 0) {
    console.log("No archived snapshots found for that domain/path/year.");
    return;
  }

  await writeOutputs(entries, outPrefix);
}

main().catch((err) => {
  console.error(err.message);
  process.exit(1);
});
